File: frontend/src/services/AdminService.js
/**
 * 管理员API服务。
 *
 * 提供学生管理、管理员管理和URL白名单管理功能。
 */
import ApiClient from './ApiClient.js'

const API_BASE = '/api'

export default class AdminService {
  constructor() {
    this.client = new ApiClient(API_BASE)
  }

  // ========== 学生管理 ==========

  /**
   * 获取学生列表。
   * @param {number} page 页码
   * @param {number} pageSize 每页数量
   * @returns {Promise<object>} 学生列表响应
   */
  async getStudents(page = 1, pageSize = 20) {
    return this.client.get(`/admin/students?page=${page}&page_size=${pageSize}`)
  }

  /**
   * 创建学生。
   * @param {string} studentId 学号
   * @param {string} password 密码
   * @param {string} displayName 显示名称
   * @returns {Promise<object>} 创建结果
   */
  async createStudent(studentId, password, displayName) {
    return this.client.post('/admin/students', {
      student_id: studentId,
      password: password,
      display_name: displayName,
    })
  }

  /**
   * 设置学生状态。
   * @param {number} userId 用户ID
   * @param {boolean} isActive 是否启用
   * @returns {Promise<object>} 操作结果
   */
  async setStudentStatus(userId, isActive) {
    return this.client.patch(`/admin/students/${userId}/status`, {
      is_active: isActive,
    })
  }

  /**
   * 删除学生。
   * @param {number} userId 用户ID
   * @returns {Promise<object>} 操作结果
   */
  async deleteStudent(userId) {
    return this.client.delete(`/admin/students/${userId}`)
  }

  // ========== 管理员管理（仅超级管理员） ==========

  /**
   * 获取管理员列表。
   * @param {number} page 页码
   * @param {number} pageSize 每页数量
   * @returns {Promise<object>} 管理员列表响应
   */
  async getAdmins(page = 1, pageSize = 20) {
    return this.client.get(`/admin/admins?page=${page}&page_size=${pageSize}`)
  }

  /**
   * 创建管理员。
   * @param {string} studentId 账号
   * @param {string} password 密码
   * @param {string} displayName 显示名称
   * @returns {Promise<object>} 创建结果
   */
  async createAdmin(studentId, password, displayName) {
    return this.client.post('/admin/admins', {
      student_id: studentId,
      password: password,
      display_name: displayName,
    })
  }

  /**
   * 设置管理员状态。
   * @param {number} userId 用户ID
   * @param {boolean} isActive 是否启用
   * @returns {Promise<object>} 操作结果
   */
  async setAdminStatus(userId, isActive) {
    return this.client.patch(`/admin/admins/${userId}/status`, {
      is_active: isActive,
    })
  }

  /**
   * 删除管理员。
   * @param {number} userId 用户ID
   * @returns {Promise<object>} 操作结果
   */
  async deleteAdmin(userId) {
    return this.client.delete(`/admin/admins/${userId}`)
  }

  // ========== 旧接口兼容 ==========

  /**
   * 获取用户列表（兼容旧接口）。
   * @param {number} page 页码
   * @param {number} pageSize 每页数量
   * @returns {Promise<object>} 用户列表响应
   */
  async getUsers(page = 1, pageSize = 20) {
    return this.client.get(`/admin/users?page=${page}&page_size=${pageSize}`)
  }

  // ========== 白名单管理 ==========

  /**
   * 获取白名单规则列表。
   * @returns {Promise<object>} 规则列表响应
   */
  async getWhitelistRules() {
    return this.client.get('/admin/whitelist')
  }

  /**
   * 创建白名单规则。
   * @param {string} ruleType 规则类型
   * @param {string} ruleValue 规则值
   * @param {string} description 描述
   * @returns {Promise<object>} 创建结果
   */
  async createWhitelistRule(ruleType, ruleValue, description = '') {
    return this.client.post('/admin/whitelist', {
      rule_type: ruleType,
      rule_value: ruleValue,
      description: description,
    })
  }

  /**
   * 更新白名单规则。
   * @param {number} ruleId 规则ID
   * @param {object} data 更新数据
   * @returns {Promise<object>} 更新结果
   */
  async updateWhitelistRule(ruleId, data) {
    return this.client.put(`/admin/whitelist/${ruleId}`, data)
  }

  /**
   * 删除白名单规则。
   * @param {number} ruleId 规则ID
   * @returns {Promise<object>} 操作结果
   */
  async deleteWhitelistRule(ruleId) {
    return this.client.delete(`/admin/whitelist/${ruleId}`)
  }

  // ========== 发件人白名单管理 ==========

  /**
   * 获取发件人白名单规则列表。
   * @returns {Promise<object>} 规则列表响应
   */
  async getSenderWhitelistRules() {
    return this.client.get('/admin/sender-whitelist')
  }

  /**
   * 创建发件人白名单规则。
   * @param {string} ruleType 规则类型
   * @param {string} ruleValue 规则值
   * @param {string} description 描述
   * @returns {Promise<object>} 创建结果
   */
  async createSenderWhitelistRule(ruleType, ruleValue, description = '') {
    return this.client.post('/admin/sender-whitelist', {
      rule_type: ruleType,
      rule_value: ruleValue,
      description: description,
    })
  }

  /**
   * 更新发件人白名单规则。
   * @param {number} ruleId 规则ID
   * @param {object} data 更新数据
   * @returns {Promise<object>} 更新结果
   */
  async updateSenderWhitelistRule(ruleId, data) {
    return this.client.put(`/admin/sender-whitelist/${ruleId}`, data)
  }

  /**
   * 删除发件人白名单规则。
   * @param {number} ruleId 规则ID
   * @returns {Promise<object>} 操作结果
   */
  async deleteSenderWhitelistRule(ruleId) {
    return this.client.delete(`/admin/sender-whitelist/${ruleId}`)
  }
}

File: backend/app/services/admin_service.py
"""管理员服务层。

提供管理员功能的业务逻辑，包括用户管理、URL白名单管理、发件人白名单管理和系统操作。
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Dict, List, Optional

from app.crud.user_crud import UserCrud
from app.crud.email_crud import EmailCrud
from app.crud.url_whitelist_crud import UrlWhitelistCrud
from app.crud.sender_whitelist_crud import SenderWhitelistCrud
from app.services.system_settings_service import SystemSettingsService

if TYPE_CHECKING:
    from app.services.phishing_detection_service import PhishingDetectionService
from app.entities.user_entity import UserEntity
from app.entities.url_whitelist_entity import UrlWhitelistEntity
from app.entities.sender_whitelist_entity import SenderWhitelistEntity
from app.schemas.admin_schema import (
    CreateUserRequest,
    CreateWhitelistRuleRequest,
    UpdateWhitelistRuleRequest,
    UserResponse,
    WhitelistRuleResponse,
    CreateSenderWhitelistRequest,
    UpdateSenderWhitelistRequest,
    SenderWhitelistResponse,
    UpdateSystemSettingsRequest,
    SystemSettingsResponse,
)


class AdminService:
    """管理员服务类。

    整合用户管理、URL白名单管理和发件人白名单管理的业务逻辑。
    """

    def __init__(
        self,
        user_crud: UserCrud,
        whitelist_crud: UrlWhitelistCrud,
        sender_whitelist_crud: SenderWhitelistCrud,
        system_settings_service: SystemSettingsService,
        email_crud: Optional[EmailCrud],
        phishing_detection_service: Optional["PhishingDetectionService"],
        logger: logging.Logger,
    ) -> None:
        """初始化管理员服务。

        Args:
            user_crud: 用户CRUD实例。
            whitelist_crud: URL白名单CRUD实例。
            sender_whitelist_crud: 发件人白名单CRUD实例。
            system_settings_service: 系统设置服务。
            email_crud: 邮件CRUD实例（用于重新检测）。
            phishing_detection_service: 钓鱼检测服务（用于重新检测）。
            logger: 日志记录器。
        """
        self._user_crud = user_crud
        self._whitelist_crud = whitelist_crud
        self._sender_whitelist_crud = sender_whitelist_crud
        self._system_settings_service = system_settings_service
        self._email_crud = email_crud
        self._phishing_detection_service = phishing_detection_service
        self._logger = logger

    # ========== 用户管理 ==========

    async def get_users(
        self, page: int = 1, page_size: int = 20
    ) -> tuple[List[UserResponse], int]:
        """获取用户列表。

        Args:
            page: 页码（从1开始）。
            page_size: 每页记录数。

        Returns:
            (用户响应列表, 总记录数) 元组。
        """
        skip = (page - 1) * page_size
        users, total = await self._user_crud.get_all_users(skip=skip, limit=page_size)

        user_responses = [
            UserResponse(
                id=user.id,
                student_id=user.student_id,
                display_name=user.display_name,
                is_active=user.is_active,
                role=user.role,
                created_at=user.created_at,
            )
            for user in users
        ]

        return user_responses, total

    async def get_students(
        self, page: int = 1, page_size: int = 20
    ) -> tuple[List[UserResponse], int]:
        """获取学生用户列表。

        Args:
            page: 页码。
            page_size: 每页记录数。

        Returns:
            (学生用户列表, 总记录数) 元组。
        """
        skip = (page - 1) * page_size
        users, total = await self._user_crud.get_students(skip=skip, limit=page_size)

        user_responses = [
            UserResponse(
                id=user.id,
                student_id=user.student_id,
                display_name=user.display_name,
                is_active=user.is_active,
                role=user.role,
                created_at=user.created_at,
            )
            for user in users
        ]

        return user_responses, total

    async def get_admins(
        self, page: int = 1, page_size: int = 20
    ) -> tuple[List[UserResponse], int]:
        """获取管理员列表（仅超级管理员可调用）。

        Args:
            page: 页码。
            page_size: 每页记录数。

        Returns:
            (管理员列表, 总记录数) 元组。
        """
        skip = (page - 1) * page_size
        users, total = await self._user_crud.get_admins(skip=skip, limit=page_size)

        user_responses = [
            UserResponse(
                id=user.id,
                student_id=user.student_id,
                display_name=user.display_name,
                is_active=user.is_active,
                role=user.role,
                created_at=user.created_at,
            )
            for user in users
        ]

        return user_responses, total

    async def create_user(
        self, request: CreateUserRequest, role: str = "user"
    ) -> Optional[UserResponse]:
        """创建新用户。

        Args:
            request: 创建用户请求。
            role: 用户角色（user/admin）。

        Returns:
            创建的用户响应，如果学号已存在返回None。
        """
        # 检查学号是否已存在
        existing = await self._user_crud.get_by_student_id(request.student_id)
        if existing:
            self._logger.warning(f"创建用户失败-学号已存在: {request.student_id}")
            return None

        user = await self._user_crud.create_with_role(
            student_id=request.student_id,
            password=request.password,
            display_name=request.display_name,
            role=role,
        )

        self._logger.info(f"创建用户成功: {request.student_id}, 角色: {role}")

        return UserResponse(
            id=user.id,
            student_id=user.student_id,
            display_name=user.display_name,
            is_active=user.is_active,
            role=user.role,
            created_at=user.created_at,
        )

    async def set_user_status(self, user_id: int, is_active: bool) -> bool:
        """设置用户启用/停用状态。

        Args:
            user_id: 用户ID。
            is_active: 是否启用。

        Returns:
            是否更新成功。
        """
        result = await self._user_crud.set_active_status(user_id, is_active)
        if result:
            self._logger.info(f"用户状态更新: user_id={user_id}, is_active={is_active}")
        return result

    async def delete_user(self, user_id: int) -> bool:
        """删除用户。

        Args:
            user_id: 用户ID。

        Returns:
            是否删除成功。
        """
        result = await self._user_crud.delete_user(user_id)
        if result:
            self._logger.info(f"删除用户: user_id={user_id}")
        return result

    # ========== 白名单管理 ==========

    async def get_whitelist_rules(self, user_id: Optional[int] = None) -> List[WhitelistRuleResponse]:
        """获取白名单规则。

        Args:
            user_id: 用户ID，如果为None则返回所有规则。

        Returns:
            白名单规则响应列表。
        """
        if user_id is not None:
            rules = await self._whitelist_crud.get_by_user_id(user_id)
        else:
            rules = await self._whitelist_crud.get_all()

        return [
            WhitelistRuleResponse(
                id=rule.id,
                rule_type=rule.rule_type,
                rule_value=rule.rule_value,
                description=rule.description,
                is_active=rule.is_active,
                created_at=rule.created_at,
            )
            for rule in rules
        ]

    async def create_whitelist_rule(
        self, request: CreateWhitelistRuleRequest, user_id: Optional[int] = None
    ) -> WhitelistRuleResponse:
        """创建白名单规则。

        Args:
            request: 创建规则请求。
            user_id: 用户ID，如果为None则创建全局规则。

        Returns:
            创建的规则响应。
        """
        rule = await self._whitelist_crud.create(
            rule_type=request.rule_type,
            rule_value=request.rule_value,
            description=request.description,
            user_id=user_id,
        )

        self._logger.info(f"创建白名单规则: {request.rule_type}:{request.rule_value}")

        return WhitelistRuleResponse(
            id=rule.id,
            rule_type=rule.rule_type,
            rule_value=rule.rule_value,
            description=rule.description,
            is_active=rule.is_active,
            created_at=rule.created_at,
        )

    async def update_whitelist_rule(
        self, rule_id: int, request: UpdateWhitelistRuleRequest
    ) -> Optional[WhitelistRuleResponse]:
        """更新白名单规则。

        Args:
            rule_id: 规则ID。
            request: 更新规则请求。

        Returns:
            更新后的规则响应，如果不存在返回None。
        """
        rule = await self._whitelist_crud.update(
            rule_id=rule_id,
            rule_type=request.rule_type,
            rule_value=request.rule_value,
            description=request.description,
            is_active=request.is_active,
        )

        if not rule:
            return None

        self._logger.info(f"更新白名单规则: rule_id={rule_id}")

        return WhitelistRuleResponse(
            id=rule.id,
            rule_type=rule.rule_type,
            rule_value=rule.rule_value,
            description=rule.description,
            is_active=rule.is_active,
            created_at=rule.created_at,
        )

    async def delete_whitelist_rule(self, rule_id: int, user_id: Optional[int] = None) -> bool:
        """删除URL白名单规则。

        Args:
            rule_id: 规则ID。
            user_id: 用户ID（用于权限验证）。

        Returns:
            是否删除成功。
        """
        result = await self._whitelist_crud.delete(rule_id, user_id)
        if result:
            self._logger.info(f"删除URL白名单规则: rule_id={rule_id}")
        return result

    # ========== 发件人白名单管理 ==========

    async def get_sender_whitelist_rules(self, user_id: Optional[int] = None) -> List[SenderWhitelistResponse]:
        """获取发件人白名单规则。

        Args:
            user_id: 用户ID，如果为None则返回所有规则。

        Returns:
            发件人白名单规则响应列表。
        """
        if user_id is not None:
            rules = await self._sender_whitelist_crud.get_by_user_id(user_id)
        else:
            rules = await self._sender_whitelist_crud.get_all()

        return [
            SenderWhitelistResponse(
                id=rule.id,
                rule_type=rule.rule_type,
                rule_value=rule.rule_value,
                description=rule.description,
                is_active=rule.is_active,
                created_at=rule.created_at,
            )
            for rule in rules
        ]

    async def create_sender_whitelist_rule(
        self, request: CreateSenderWhitelistRequest, user_id: Optional[int] = None
    ) -> SenderWhitelistResponse:
        """创建发件人白名单规则。

        Args:
            request: 创建规则请求。
            user_id: 用户ID，如果为None则创建全局规则。

        Returns:
            创建的规则响应。
        """
        rule = await self._sender_whitelist_crud.create(
            rule_type=request.rule_type,
            rule_value=request.rule_value,
            description=request.description,
            user_id=user_id,
        )

        self._logger.info(
            f"创建发件人白名单规则: {request.rule_type}:{request.rule_value}"
        )

        return SenderWhitelistResponse(
            id=rule.id,
            rule_type=rule.rule_type,
            rule_value=rule.rule_value,
            description=rule.description,
            is_active=rule.is_active,
            created_at=rule.created_at,
        )

    async def update_sender_whitelist_rule(
        self, rule_id: int, request: UpdateSenderWhitelistRequest
    ) -> Optional[SenderWhitelistResponse]:
        """更新发件人白名单规则。

        Args:
            rule_id: 规则ID。
            request: 更新规则请求。

        Returns:
            更新后的规则响应，如果不存在返回None。
        """
        rule = await self._sender_whitelist_crud.update(
            rule_id=rule_id,
            rule_type=request.rule_type,
            rule_value=request.rule_value,
            description=request.description,
            is_active=request.is_active,
        )

        if not rule:
            return None

        self._logger.info(f"更新发件人白名单规则: rule_id={rule_id}")

        return SenderWhitelistResponse(
            id=rule.id,
            rule_type=rule.rule_type,
            rule_value=rule.rule_value,
            description=rule.description,
            is_active=rule.is_active,
            created_at=rule.created_at,
        )

    async def delete_sender_whitelist_rule(self, rule_id: int, user_id: Optional[int] = None) -> bool:
        """删除发件人白名单规则。

        Args:
            rule_id: 规则ID。
            user_id: 用户ID（用于权限验证）。

        Returns:
            是否删除成功。
        """
        result = await self._sender_whitelist_crud.delete(rule_id, user_id)
        if result:
            self._logger.info(f"删除发件人白名单规则: rule_id={rule_id}")
        return result

    # ========== 系统设置管理 ==========

    async def get_system_settings(self) -> SystemSettingsResponse:
        """获取系统设置。

        Returns:
            系统设置响应。
        """
        settings = await self._system_settings_service.get_settings()
        return SystemSettingsResponse(
            enable_long_url_detection=settings.enable_long_url_detection,
            enable_rule_based_detection=settings.enable_rule_based_detection,
            created_at=settings.created_at,
            updated_at=settings.updated_at,
        )

    async def update_system_settings(
        self, request: UpdateSystemSettingsRequest
    ) -> SystemSettingsResponse:
        """更新系统设置。

        Args:
            request: 更新系统设置请求。

        Returns:
            更新后的系统设置响应。
        """
        settings = await self._system_settings_service.update_settings(
            enable_long_url_detection=request.enable_long_url_detection,
            enable_rule_based_detection=request.enable_rule_based_detection,
        )
        return SystemSettingsResponse(
            enable_long_url_detection=settings.enable_long_url_detection,
            enable_rule_based_detection=settings.enable_rule_based_detection,
            created_at=settings.created_at,
            updated_at=settings.updated_at,
        )